### PR TITLE
Enemy detection

### PR DESCRIPTION
**Pull request by robertecurtin**
This is untested, but it _should_ tie together all the minimum pieces required for the strategist and the charge / seek / line avoidance tactics.

I've confirmed that the distance sensors are working to some extent. I currently have it set up so that the left sensor (from the robot's perspective) is plugged into A3, and the right sensor is plugged into A2. I was not able to read from the other sensors for some reason.

Here's a terrible diagram of the minimum system:
diagram

### CODE REVIEW

File: src/app/common/meta-sensing/detect_enemy.c [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+void detect_enemy_init(
+  detect_enemy_t* self,
+  i_tiny_key_value_store_t* key_value_store,
+  const detect_enemy_keys_t* keys) {
+  self->key_value_store = key_value_store;
+  self->keys = keys;
+
+  tiny_event_subscription_init(&self->on_change_subscription, self, data_changed);
+  tiny_event_subscribe(tiny_key_value_store_on_change(key_value_store), &self->on_change_subscription);
+}
```
Comment (ryanplusplus) [inline src/app/common/meta-sensing/detect_enemy.c]: Editor settings are bonk

File: src/app/starduino-zumo/line_sensors_plugin.c [hunk 2]
```diff
@@ -54,15 +54,20 @@ static void end_capture(void) {
...
 
 static void measure(line_sensors_plugin_t* self) {
+  bool left_line_detected;
+  bool right_line_detected;
   bool line_detected;
 
   uint16_t left_count = (TIM1->CCR3H << 8) + TIM1->CCR3L;
-  line_detected = left_count < line_threshold;
-  tiny_key_value_store_write(self->key_value_store, key_left_line_detected, &line_detected);
+  left_line_detected = left_count < line_threshold;
```
Comment (ryanplusplus) [inline src/app/starduino-zumo/line_sensors_plugin.c]: Why not declare here?

File: src/app/starduino-zumo/line_sensors_plugin.c [hunk 3]
```diff
@@ -54,15 +54,20 @@ static void end_capture(void) {
...
+  tiny_key_value_store_write(self->key_value_store, key_left_line_detected, &left_line_detected);
 
   uint16_t right_count = (TIM1->CCR4H << 8) + TIM1->CCR4L;
-  line_detected = right_count < line_threshold;
-  tiny_key_value_store_write(self->key_value_store, key_right_line_detected, &line_detected);
+  right_line_detected = right_count < line_threshold;
+  tiny_key_value_store_write(self->key_value_store, key_right_line_detected, &right_line_detected);
+
+  line_detected = left_line_detected || right_line_detected;
```
Comment (ryanplusplus) [inline src/app/starduino-zumo/line_sensors_plugin.c]: Why not declare here?

File: src/app/common/meta-sensing/detect_enemy.c [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+enum {
+  threshold = 20
+};
+
+static void data_changed(void* context, const void* _args) {
+  reinterpret(self, context, detect_enemy_t*);
+  reinterpret(args, _args, const tiny_key_value_store_on_change_args_t*);
+
+  if(args->key == self->keys->sensor_distance) {
+    reinterpret(distance, args->value, distance_in_cm_t*);
```
Comment (ryanplusplus) [inline src/app/common/meta-sensing/detect_enemy.c]: Lost your const

File: src/app/common/meta-sensing/detect_enemy.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+#ifndef detect_enemy_h
+#define detect_enemy_h
+
+#include "i_tiny_key_value_store.h"
+
+typedef struct {
+  tiny_key_value_store_key_t sensor_distance;
+  tiny_key_value_store_key_t sensor_active;
```
Comment (ryanplusplus) [inline src/app/common/meta-sensing/detect_enemy.h]: This was a bit unclear to me. How about something along the lines of "enemy detected"?

File: src/app/common/meta-sensing/detect_enemy.c [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  threshold = 30
+};
+
+static void data_changed(void* context, const void* _args) {
+  reinterpret(self, context, detect_enemy_t*);
+  reinterpret(args, _args, const tiny_key_value_store_on_change_args_t*);
+
+  if(args->key == self->keys->sensor_distance) {
+    reinterpret(distance, args->value, const distance_in_cm_t*);
+    bool is_active = *distance <= threshold;
```
Comment (robertecurtin) [inline src/app/common/meta-sensing/detect_enemy.c]:
```suggestion

    bool is_active = *distance <= threshold;
```

**Pull request merged**